File: flutter/mobile/lib/models/product.dart
import 'brand.dart';

class Product {
  final int id;
  final String name;
  final double price;
  final int quantityInStock;

  // ⭐ RATING
  final double avgRating;
  final int totalReviews; // dùng cái này cho UI

  final List<String> imageUrls;
  final Brand? brand;

  // ===== SPEC =====
  final String ram;
  final String storage;
  final String chipset;
  final String camera;
  final String battery;
  final String screenSize;
  final String color;
  final String description;
  final String os;

  Product({
    required this.id,
    required this.name,
    required this.price,
    required this.quantityInStock,
    required this.avgRating,
    required this.totalReviews,
    required this.imageUrls,
    this.brand,
    required this.ram,
    required this.storage,
    required this.chipset,
    required this.camera,
    required this.battery,
    required this.screenSize,
    required this.color,
    required this.description,
    required this.os,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json["id"],
      name: json["name"],
      price: (json["price"] as num).toDouble(),
      quantityInStock: json["quantityInStock"] ?? 0,

      // ⭐ MAP RATING (AN TOÀN)
      avgRating: (json["avgRating"] ?? 0).toDouble(),
      totalReviews: json["totalReviews"] ?? 0,

      imageUrls:
          (json["imageUrls"] as List?)
              ?.map(
                (e) => e
                    .toString()
                    .replaceAll("localhost", "10.0.2.2")
                    .replaceAll("\\", "/"),
              )
              .toList() ??
          [],

      brand: json["brand"] != null ? Brand.fromJson(json["brand"]) : null,

      // ===== MAP SPEC =====
      ram: json["ram"] ?? "",
      storage: json["storage"] ?? "",
      chipset: json["chipset"] ?? "",
      camera: json["camera"] ?? "",
      battery: json["battery"] ?? "",
      screenSize: json["screenSize"] ?? "",
      color: json["color"] ?? "",
      description: json["description"] ?? "",
      os: json["os"] ?? "",
    );
  }
}

File: backend/tripplaner/src/main/java/kj002/tripplaner/models/Product.java
package kj002.tripplaner.models;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "products")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "brand_id", nullable = false)
    @JsonManagedReference
    private Brand brand;

    @Column(nullable = false)
    private String name;
    @Column(nullable = false)
    private double price;
    private Integer quantityInStock;
    private String screenSize;
    private String ram;
    private String storage;
    private String chipset;
    private String camera;
    private String battery;
    private String os;
    private String color;
    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    private ProductStatus status = ProductStatus.ACTIVE;

    private LocalDateTime createdAt = LocalDateTime.now();

    @OneToMany(
            cascade = CascadeType.ALL,
            mappedBy = "product",
            orphanRemoval = true
    )
    private List<ProductImage> images=new ArrayList<>();
}
